### PR TITLE
TIG-1209: Don't expose DefaultRandom to Actors

### PR DESCRIPTION
**Pull request by rtimmons**
Okay so Actors still *can* use `DefaultRandom` (and one still does for its own reasons) but the point is that now Actors can just call `context.createDocumentGenerator(id,"Document")` or somesuch without having to worry about the RNG details.

I tried to get cute to enable `context.get<DocumentGenerator>()` but that didn't work out - partially because e.g. `CrudActor` does want to construct DocumentGenerators from `YAML::Node` not just via `ConfigNode`.

### CODE REVIEW

File: src/cast_core/src/CommitLatency.cpp [hunk 1]
```diff
@@ -178,7 +179,7 @@ void CommitLatency::run() {
 
 CommitLatency::CommitLatency(genny::ActorContext& context)
     : Actor(context),
-      _rng{context.workload().createRNG()},
+      _rng{context.workload().getRNGForActor(CommitLatency::id())},
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/cast_core/src/CommitLatency.cpp]: I'm down for renaming to `getRNGForThread` but renaming `ActorId` is a bigger change that I'd want to consider.

File: src/gennylib/src/context.cpp [hunk 2]
```diff
@@ -98,12 +98,19 @@ v1::GlobalRateLimiter* WorkloadContext::getRateLimiter(const std::string& name,
...
-    if (_done) {
-        throw InvalidConfigurationException(
-            "Tried to create a random number generator after construction");
+DefaultRandom& WorkloadContext::getRNGForActor(ActorId id) {
+    if (this->isDone()) {
+        BOOST_THROW_EXCEPTION(std::logic_error("Cannot create RNGs after setup"));
     }
-    return DefaultRandom{_rng()};
+    if (auto rng = _rngRegistry.find(id); rng == _rngRegistry.end()) {
+        auto [it, success] = _rngRegistry.try_emplace(id, DefaultRandom{_rng()});
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/gennylib/src/context.cpp]: Yup good call - changed

File: src/gennylib/include/gennylib/context.hpp [hunk 3]
```diff
@@ -406,9 +522,56 @@ class PhaseContext final : public v1::ConfigNode {
...
+        msg << "[";
+        msg << pathOrNode;
+        (msg << ... << args) << "]";
+        msg << " after workload setup completed.";
+        BOOST_THROW_EXCEPTION(std::logic_error(msg.str()));
+    }
+
+    DefaultRandom& rng = workloadContext.getRNGForActor(id);
+
+    if constexpr (IsLooselyConvertible<PathOrNode, YAML::Node>::value) {
```
[the remark this answers is not in the document]
Comment (rtimmons) [reply src/gennylib/include/gennylib/context.hpp]: *Fortunately* there's no implicit `YAML::Node(string)` ctor so this handles strings correctly. 
I'm not sure what you mean by "Is it better to ask the caller to do line 566?" - can  you elaborate?

(fwiw I'm not in love with users *ever* being exposed to `YAML::Node` I'd much rather only let them interact with `ConfigNode` which may or may not make this situation a bit easier...but that's a bigger change for another day....)

**Comment by rtimmons** [comment]
Thanks for review @guoyr. Implemented feedback ready for another check whenever you have a minute 👀

I have no idea why the Orchestrator perf test is reliably failing on arch this change should have nothing to do with it. I'm going to take a deeper look into it.

File: src/gennylib/include/gennylib/context.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (guoyr) [reply src/gennylib/include/gennylib/context.hpp]: Ah ok I think if we don't want users to interact with bare `YAML::Node` then my concern goes away.

> I'm not sure what you mean by "Is it better to ask the caller to do line 566?" - can you elaborate?

I was wondering if we can get rid of the `else` clause and have the user pass in the result of `configNode.get<>` instead.

File: src/value_generators/include/value_generators/DefaultRandom.hpp [hunk 4]
```diff
@@ -44,10 +44,11 @@ class Random {
      */
     explicit Random(result_type seed = 6514393) : _rng(seed) {}
 
-    // No moves or copies.
-    Random(Random&&) = delete;
-    Random& operator=(Random&&) = delete;
+    // Moves are okay
+    Random(Random&&) noexcept = default;
```
Comment (guoyr) [inline src/value_generators/include/value_generators/DefaultRandom.hpp]: No need for these ctors now

**Comment by guoyr** [comment]
> I have no idea why the Orchestrator perf test is reliably failing on arch this change should have nothing to do with it. I'm going to take a deeper look into it.

Yeah It's been failing for me too. Definitely not (only) related to this PR.

**Pull request merged**